### ISSUE TITLE
{column}_colors must be validated

### ISSUE THREAD

**Issue by brianraymor** [chanzuckerberg/single-cell-curation#513]
**Note**: Per Arathi's direction in [#single-cell-planning](https://czi-sci.slack.com/archives/C04PHE5TV9B/p1683819000421289):

 **_Any validation code against the CXG schema will also be owned by the Data Visualization team._**

# Design

### {column}_colors

<table><tbody>
  <tr>
    <th>Key</th>
      <td>
        {column}_colors where {column} MUST be the name of a <code>category</code> data type column in <code>obs</code> that<br> is annotated by the data submitter or curator. The following columns that are annotated by CELLxGENE<br> Discover MUST NOT be specified as {column}:<br><br>
      <ul>
        <li>assay</li>
        <li>cell_type</li>
        <li>development_stage</li>
        <li>disease</li>
        <li>organism</li>
        <li>self_reported_ethnicity</li>
        <li>sex</li>
        <li>tissue</li>       
      </ul><br>
      Instead annotate {column}_ontology_term_id_colors for these columns such as <code>assay_ontology_term_id</code>.<br><br>
    </td>
  </tr>
  <tr>
    <th>Annotator</th>
    <td>Curator</td>
  </tr>
  <tr>
    <th>Value</th>
      <td>
        <code>numpy.ndarray</code>. This MUST be a 1-D array of shape <code>(, c)</code>, where <code>c</code> is greater than or equal to the<br> number of unique categories in the {column} as calculated by:<br><br>
           <samp>len(anndata.obs.{column}.unique())</samp><br><br>
        The color code at the Nth position in the <code>ndarray</code> corresponds to the Nth category of <samp>anndata.obs.{column}.unique()</samp>.<br><br>For example, if <code>cell_type_ontology_term_id</code> includes two unique categories:<br><br>
        <samp>anndata.obs.cell_type_ontology_term_id.unique()</samp><br><br>
        <samp>['CL:0000057', 'CL:0000115']<br>Categories (2, object): ['CL:0000057', 'CL:0000115']</samp><br><br>then <code>cell-type_ontology_term_id_colors</code> MUST contain two or more colors such as:<br><br>
        <samp>['aqua' 'blueviolet']</samp><br><br>where <code>'aqua'</code> is the color assigned to <code>'CL:0000057'</code> and <code>'blueviolet'</code> is the color assigned to<br> <code>'CL:0000115'</code>.<br><br>All elements in the <code>ndarray</code> MUST use the same color model, limited to:<br><br>
          <table>
          <thead>
            <tr>
              <th>Color Model</th>
              <th>Element Format</th>
            </tr>
          </thead><tbody>
            <tr>
              <td>
              <a
              href="https://www.w3.org/TR/css-color-4/#named-colors"
              ><i>Named Colors </i>
              </a>
            </td>
              <td><code>str</code>. MUST be a case-insensitive CSS4 color name with no spaces such as<br> <code>"aliceblue"</code>
            </td>
            </tr>
            <tr>
             <td>
              <a
              href="https://www.w3.org/TR/css-color-4/#hex-notation"
              ><i>Hex Triplet</i>
              </a>
            </td>
              <td><code>str</code>. MUST start with <code>"#"</code> immediately followed by six case-insensitive hexadecimal<br> characters as in <code>"#08c0ff"</code></td>
            </tr>
          </tbody></table>
        </td>
    </tr>
</tbody></table>
<br>

**Comment by brianraymor** [comment]
@atarashansky - I updated the top-level summary comments with the requirements. Please contact @jahilton and @jychien when they can test the changes.

**Comment by jahilton** [comment]
I'm getting these error sfor all tests...

> ERROR: Annotated categorical field donor_id must have at least 44 color options in uns[donor_id_colors]. Found: []
ERROR: Annotated categorical field suspension_type must have at least 1 color options in uns[suspension_type_colors]. Found: []
ERROR: Annotated categorical field tissue_type must have at least 1 color options in uns[tissue_type_colors]. Found: []

Confirmed that in each case, there is no donor_id_colors, suspension_type_colors, or tissue_type_colors in uns.

**Comment by joyceyan** [comment]
@jahilton Thanks! I think I misunderstood the expected behavior. Just to confirm, if `{column}_colors` is not defined in `uns`, then that should still pass validation right? We should only fail if the expected number of colors is incorrect or if we have invalid colors?

**Comment by jahilton** [comment]
Correct, we're not expecting/requiring any {column}_colors to be present, only validating the ones that are present.

**Comment by brianraymor** [comment]
Hmm. 

This field is in a schema section that states:

> ​Curators MAY also annotate the following optional keys and values in uns. If the key is present, then its value MUST NOT be empty.​

Maybe it would be helpful to embed that information per field for clarity.

**Comment by jahilton** [comment]
What we're expected to pass validation now passes validation.
Unfortunately, so does everything we expect to fail - [QA ntbk](https://github.com/Lattice-Data/lattice-tools/blob/CXG290-schema4validator-testing/cellxgene_resources/cxg_4.0.0_testing/513_colors.ipynb).

- length of colors array is less than obs counterpart
- colors array that is empty np array
- colors array that is np.nan
- colors array that is empty strings
- colors as string or dictionary
- colors for schema fields
- colors does not have obs counterpart
- colors is mixed named & hex
- colors is non-named or hex
- colors counterpart in obs is boolean, float, or int

**Comment by joyceyan** [comment]
@jahilton Just to clarify some of the requirements here:
1. For colors that don't have an obs counterpart, should these raise a warning or an error?
2. My understanding was that the colors counterpart in obs had to be a `categorical` type. Do we want to further verify that it should be a `categorical` type of `boolean`, `float` or `int`? If so, I think we should make that clear in the schema definition.
3.

**Comment by signechambers1** [comment]
@joyceyan i defer to @jahilton's expertise on #1 but happy to weigh in if helpful.

Re: #2 You are correct, the colors counterpart in obs should be a categorical data type. colors defined for fields that are boolean, float or int should fail validation.

**Comment by jahilton** [comment]
>For colors that don't have an obs counterpart, should these raise a warning or an error?

The schema uses **MUST** in this statement, which equates to an error (and failed validation)...
>{column}_colors where {column} MUST be the name of a `category` data type column in `obs`

**Comment by brianraymor** [comment]
> My understanding was that the colors counterpart in obs had to be a categorical type. Do we want to further verify that it should be a categorical type of boolean, float or int? If so, I think we should make that clear in the schema definition.

I believe that the schema is precise:

> {column}_colors where {column} MUST be the name of a category data type column in obs

It simply needs to be a **category data type**.  I'd note that most of the schema fields are defined as categoricals with `str` categories such as `assay_ontology_term_id`. 

For example, `adata.obs.dtypes` returns:

```
BICCN_cluster_id                             float64
QC                                          category
BICCN_cluster_label                         category
BICCN_subclass_label                        category
BICCN_class_label                           category
cluster_color                               category
size                                         float64
temp_class_label                            category
BICCN_ontology_term_id                      category
assay_ontology_term_id                      category
disease_ontology_term_id                    category
tissue_ontology_term_id                     category
BICCN_project                               category
cell_type_ontology_term_id                  category
self_reported_ethnicity_ontology_term_id    category
development_stage_ontology_term_id          category
sex_ontology_term_id                        category
is_primary_data                                 bool
organism_ontology_term_id                   category
donor_id                                    category
suspension_type                             category
tissue_type                                 category
cell_type                                   category
assay                                       category
disease                                     category
organism                                    category
sex                                         category
tissue                                      category
self_reported_ethnicity                     category
development_stage                           category
dtype: object
```

**Comment by atarashansky** [comment]
@brianraymor
Unrelated, but it would be nice to enforce "bool" data types to be categorical

**Comment by brianraymor** [comment]
> Unrelated, but it would be nice to enforce "bool" data types to be categorical

We could add that requirement to `is_primary_data`.

However, when I review the CXG code, its source of truth is whether a name is present in `adata.obs.keys()` and does not require a categorical?

https://github.com/chanzuckerberg/single-cell-data-portal/blob/d042bdadb17f2fa389f537ed26eb429698350596/backend/common/utils/color_conversion_utils.py#L224

```
    cxg_colors = dict()
    color_key_suffix = "_colors"
    for uns_key in adata.uns:
        # find uns array that describes colors for a category
        if not uns_key.endswith(color_key_suffix):
            continue

        # check to see if we actually have observations for that category
        category_name = uns_key[: -len(color_key_suffix)]
        if category_name not in adata.obs.keys():
            continue

        # create the cellxgene color entry for this category
        cxg_colors[category_name] = dict(
            zip(adata.obs[category_name].cat.categories, [convert_color_to_hex_format(c) for c in adata.uns[uns_key]])
        )
```


CC: @nayib-jose-gloria

**Comment by joyceyan** [comment]
This is ready for a second round of QA now. I think I misunderstood earlier, but I believe we are all aligned now that for a color field in `uns[{column}_colors]` must have a corresponding `category` data type column in `obs`.

The CXG code does not appear to require that a defined user color has a corresponding `obs` column, but my understanding is that we want most of the validation logic to live here in the CLI, so I'm not sure if it's a problem that we don't validate this in the CXG converter.

**Comment by jahilton** [comment]
For any `uns[{column}_colors]`, I am getting this error resulting in failed validation...
`ERROR: Colors field uns[BICCN_class_label_colors] does not have a corresponding categorical field in obs`
despite `BICCN_class_label` being present in `obs` and `adata.obs[BICCN_class_label].dtype` == `category`
[QA notebook](https://github.com/Lattice-Data/lattice-tools/blob/CXG290-schema4validator-testing/cellxgene_resources/cxg_4.0.0_testing/513_colors.ipynb) shows the unexpected behavior for AnnData objects expected to be valid

**Comment by joyceyan** [comment]
@jahilton Hm interesting, the equivalent tests I have on my end seem to pass, but it looks like we're probably setting up our sample data a bit differently. It appears there's actually two ways you can mark a column as "categorical" - you can set the column definition's type to be `categorical` as well as the column's dtype = `category`. The column definition is stored within the schema definition, which is separate from the `adata` object. It seems like we have separate validation that verifies that if you set the column definition's type to be `categorical`, we require that the column's dtype is also `category`, but not the other way around. This seems to imply that column definitions in the schema are optional, but when provided, are required to match the corresponding column's dtype.

For this check, the current implementation is only looking at the column definition type, and not at the column's dtype directly. In general, is the column dtype (`adata.obs["BICCN_class_label"].dtype`) a more reliable and the column definition (`schema_def.obs["BICCN_class_label"]["type"]`) an optional thing people can provide? I'm trying to understand which of the following options makes the most sense implementation wise:
1. Verify that column definition = categorical
2. Verify that column dtype = category
3. Verify that both column definition = categorical and column dtype = category

The current implementation is (1) which seems to be almost certainly wrong, but it's not clear to me whether (2) or (3) is better.

**Comment by Bento007** [comment]
Option 3 because the `schema_def` dictates what the format should be because it's what we expect. This might mean we need to modify the `schema_def` more `dtypes` should be supported.

**Comment by nayib-jose-gloria** [comment]
@joyceyan @Bento007 

tldr: We should only verify column dtype = category

longer answer: the schema definition is not an input from users; it's dev-defined for a particular release. It's used by the validator to determine what validations to run on which columns. So, if a field name is annotated with 'categorical' in the [schema def](https://github.com/chanzuckerberg/single-cell-curation/blob/main/cellxgene_schema_cli/cellxgene_schema/schema_definitions/schema_definition.yaml), then we run the validation rule 'Verify that column dtype = category' for that column. This is for columns pre-defined in the documented [schema definition](https://github.com/chanzuckerberg/single-cell-curation/blob/main/schema/4.0.0/schema.md#tissue_type).

if I understand correctly, the columns being validated here could be called literally anything so long as they have a corresponding `_colors` column in the uns, so they won't be in the schema definition and the schema def type will always be` != categorical`. We could rework the schema definition + validator yaml parsing to accommodate this use-case, but that seems like a larger refactor

**Comment by joyceyan** [comment]
@nayib-jose-gloria I see - that makes sense, I appreciate the clarification!

**Comment by joyceyan** [comment]
@jahilton I've updated the underlying implementation to look at the column's dtype, rather than checking the schema definition's type. I believe this should fix the issue you were seeing, but let me know if you still see issues in QA.

**Comment by jahilton** [comment]
Now getting mostly 'Colors...must be string' errors like...
`ERROR: Colors in uns[BICCN_class_label_colors] must be strings. Found: ['slategrey' 'firebrick' 'mediumorchid']`
...and I'm not sure why

A couple notes aside from that for the ERROR messages in the expected invalid cases

- Currently, if a schema _portal_ field is used in colors: `ERROR: Colors field uns[cell_type_colors] does not have a corresponding categorical field in obs`. Would be great to add on: `Annotate {column}_ontology_term_id_colors for this`
- When a corresponding obs field is present but it is not categorical, current is: `ERROR: Colors field uns[{column}_colors] does not have a corresponding categorical field in obs`. Would help the curator diagnose if it was appended with: `{column} is present but is dtype {}`

**Comment by joyceyan** [comment]
@jahilton The check we use for "does this numpy array contain strings" is we check if `numpy.issubdtype(array.dtype, numpy.character)`. Looking at the output of the Lattice notebook, the issue appears to be isolated to test cases that are set up with `mcolors.CSS4_COLORS.keys()`. When I try to recreate your test setup with something that looks like this
```
validator.adata.uns["suspension_type_colors"] = numpy.array(random.sample(mcolors.CSS4_COLORS.keys(), 2))
```
It seems to pass validation, at least on my machine, even in two different virtual Python environments. I do, however, get two different results when I run something like this in two different environments I have:
```
test_array = np.array(random.sample(mcolors.CSS4_COLORS.keys(), 3))
print(test_array.dtype)
print(test_array.dtype.name)
```
In one environment I get a `<U10` dtype (named `str320`), and in another I get a `<U4` dtype (named `str448`). Either way, both seem to pass `numpy.issubdtype(array.dtype, numpy.character)`, both when I run the validation as well as when I run something like this:
```
print(numpy.issubdtype('<U10', numpy.character))
print(numpy.issubdtype('<U14', numpy.character))
```

My suspicion is that whatever version of python and/or matlab and/or numpy you're running is generating `mcolors.CSS4_COLORS.keys()` as some numpy dtype that isn't `str320` or `str448` and isn't a subdtype of `numpy.character`.

I would probably expect people submitting datasets to just be hard-coding a list of colors rather than programmatically generating them, so if the issue is just updating the Lattice test setup to make sure we're generating strings, I think updating the test setup to look something like this would do it. The `U4` data type is a `str128`.
```
np.array(random.sample(mcolors.CSS4_COLORS.keys(), 3), dtype='U4')
```

If it's important to support the tests as they're currently written, it would be helpful for me to know what version of Python, numpy, and matlab you're running, as well as the `dtype` and `dtype.name` of the arrays that are generated so that I can try to reproduce the issue on my side.

**Comment by jahilton** [comment]
@joyceyan are you printing the dtype/dtype.name before or after saving the AnnData?

I can set the uns.*_colors field and I get <U7 & str224. Then I write AnnData, read it in, and do the same evaluation of the same uns.*_colors field and I get object & object.

I suspect that writing AnnData is converting the array in a way that makes it impossible to create a valid *_colors field as the validator is currently written. If you are able to create a valid *_colors list, please let me know how.

**Comment by jahilton** [comment]
For "real world" examples that should pass validation, [Tabula Sapiens datasets](https://cellxgene.cziscience.com/collections/e5f58829-1a66-40b5-a624-9046778e74f5) have several *_colors defined.

**Comment by joyceyan** [comment]
I see - the tests I have as written follow the formula of creating the AnnData object first, and then adding things to `adata.uns[{column}_colors]`. I've been able to set up a test where I create the `_colors` list in the AnnData object directly, so I can reproduce the same error you're getting where it appears to convert the array into objects. I'll keep you posted once I figure out how to update the validator to not flag these as "not strings".

**Comment by joyceyan** [comment]
@jahilton I've updated how we detect strings in the AnnData, and I believe this shouldn't be a problem anymore. Let me know how things look on your end!

**Comment by jahilton** [comment]
Ran the cases in the [QA notebook](https://github.com/Lattice-Data/lattice-tools/blob/CXG290-schema4validator-testing/cellxgene_resources/cxg_4.0.0_testing/513_colors.ipynb) that we expect to be valid...

- Some CSS4 colors seem to be invalid `ERROR: Colors in uns[<field>_colors] must be either all hex colors or all CSS4 named colors.`
  - CSS4 colors in the tests are randomly selected from matplotlib.colors.CSS4_COLORS.keys()
  - At least one of these is invalid...`['azure' 'beige' 'coral' 'darkslategray' 'deeppink' 'deepskyblue' 'floralwhite' 'ivory' 'lightgrey' 'lightslategray' 'powderblue' 'purple' 'skyblue' 'slategrey' 'tomato' 'violet' 'yellowgreen']`
- The validator appears to be counting the number of _unique_ colors defined, rather than the total colors defined...
```
ERROR: Annotated categorical field BICCN_class_label must have at least 3 color options in uns[BICCN_class_label_colors]. Found: ['silver' 'snow']
ERROR: Annotated categorical field cell_type_ontology_term_id must have at least 6 color options in uns[cell_type_ontology_term_id_colors]. Found: ['#8B4513' '#CD853F' '#E6E6FA']
```
But the lists in uns are actually of sufficient length -  `['silver' 'snow' 'silver' 'snow']` and `['#8B4513' '#E6E6FA' '#CD853F' '#8B4513' '#E6E6FA' '#CD853F']`. The schema does not specify _unique_ in reference to the defined colors in uns

**Comment by joyceyan** [comment]
For the colors, I hard-coded a list of CSS4 named colors from some list that appears to be outdated (or incorrect) and doesn't include `lightgrey`. I can update the validator to pull from `matplotlib.colors.CSS4_COLORS` which should make things consistent.

Understood on the fact that colors don't need to be unique - it seems a bit odd to me to allow that, but that can easily be fixed as well.

**Comment by joyceyan** [comment]
Even if we do want to allow duplicate colors to pass the validator, it might make sense to raise a warning.

**Comment by jahilton** [comment]
Use case: contributor wants two categories colored the same and doesn't want to create a new obs field to accomplish that grouping. Example: 5 assays in a dataset - Smart-seq, sci-RNA-seq, 10x 3' v2, 10x 3' v3, 10x 5' v1. They may want the 10x 3' v2/v3 to be colored the same, but still retain the more specific information.

Unless Explorer has a specific reason to block that, then I think we should also allow it.

**Comment by nayib-jose-gloria** [comment]
gotcha--thanks for the explanation!

**Comment by joyceyan** [comment]
@jahilton The validator has been updated to allow duplicate colors and match the CSS4 colors that `matplotlib` provides. Let me know how things look on your end.

**Comment by nayib-jose-gloria** [comment]
@jahilton the change involved adding a new dependency (matplotlib) btw, heads up to pip install without that` no-deps `flag

**Comment by jahilton** [comment]
Behavior is now as expected for all tests, valid & invalid.
Only outstanding thing is this [comment from above](https://github.com/chanzuckerberg/single-cell-curation/issues/513#issuecomment-1751485365)

- Currently, if a schema _portal_ field is used in colors: `ERROR: Colors field uns[cell_type_colors] does not have a corresponding categorical field in obs.` Would be great to add on: `Annotate {column}_ontology_term_id_colors instead.`

**Comment by joyceyan** [comment]
@jahilton Thanks for flagging - I missed that in my earlier update. The validator should now include "Annotate {column}_ontology_term_id_colors instead" if the column name is any of:
- assay
- cell_type
- development_stage
- disease
- organism
- self_reported_ethnicity
- sex
- tissue

**Comment by jahilton** [comment]
LGTM

**Comment by joyceyan** [comment]
Awesome! Closing this ticket out now 🎉
